<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Viktor Petukhov" />

<meta name="date" content="2021-11-30" />

<title>Preprocess the MS dataset</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">cacoaAnalysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Preprocess the MS dataset</h1>
<h4 class="author"><a href="https://github.com/VPetukhov/">Viktor Petukhov</a></h4>
<h4 class="date">2021-11-30</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-11-30
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>cacoaAnalysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it's best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20211123code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20211123)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20211123code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20211123)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong3f1371c"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> 3f1371c </a>
</p>
</div>
<div id="strongRepositoryversionstrong3f1371c" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version 3f1371c. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/de_figure_ep.nb.html
    Ignored:    analysis/figure/
    Ignored:    analysis/preprocess.nb.html
    Ignored:    analysis/site_libs/
    Ignored:    data/ASD/
    Ignored:    data/AZ/
    Ignored:    data/EP/
    Ignored:    data/MS/
    Ignored:    data/PF/
    Ignored:    data/SCC/
    Ignored:    man/

Untracked files:
    Untracked:  analysis/cluster_free_de_pf.Rmd
    Untracked:  analysis/de_figure_ep.Rmd
    Untracked:  analysis/fig_cluster_free_de.Rmd
    Untracked:  cache/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/preprocess.Rmd</code>) and HTML (<code>docs/preprocess.html</code>) files. If you've configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
3f1371c
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
<td>
Cleaner preprocessing
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
1d802ad
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
<td>
Force rebuild in preprocess.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
2212721
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
<td>
Small fix in SCC preprocessing
</td>
</tr>
<tr>
<td>
html
</td>
<td>
3085bf5
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
<td>
Build preprocessing.html
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
70d2759
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
<td>
Preprocessing notebook
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="libraries" class="section level2">
<h2>Libraries</h2>
<pre class="r"><code>library(tidyverse)
library(magrittr)
library(sccore)
library(pagoda2)
library(conos)
library(dataorganizer)
library(Matrix)
library(reticulate)

devtools::load_all()

N_CORES &lt;- 45
FORCE &lt;- TRUE</code></pre>
<pre class="r"><code>virtualenv_create(&quot;r-scrublet&quot;)</code></pre>
<pre><code>virtualenv: r-scrublet</code></pre>
<pre class="r"><code>virtualenv_install(&quot;r-scrublet&quot;, c(&quot;scrublet&quot;, &quot;matplotlib&quot;))</code></pre>
<pre><code>Using virtual environment &#39;r-scrublet&#39; ...</code></pre>
<pre class="r"><code>use_virtualenv(&quot;r-scrublet&quot;)
mpl &lt;- import(&quot;matplotlib&quot;)
mpl$use(&#39;Agg&#39;) # Otherwise it shows Qt error in RStudio
scrublet &lt;- import(&quot;scrublet&quot;)

estimateDoubletInfo &lt;- function(mats, progress=FALSE) {
  dub.info &lt;- sccore::plapply(mats, function(m) {
    suppressMessages(scrublet$Scrublet(t(m), random_state=as.integer(42))$scrub_doublets()) %&gt;%
      lapply(setNames, colnames(m))
  }, n.cores=1, progress=progress)

  lapply(c(scores=1, mask=2), function(i) {
    lapply(dub.info, `[[`, i) %&gt;% unname() %&gt;% unlist()
  })
}</code></pre>
</div>
<div id="alzheimer" class="section level2">
<h2>Alzheimer</h2>
<p>Cells are already filtered by mit. fraction, and doublets are removed. We filter only by minimum 500 UMIs per cell and scrublet scores.</p>
<pre class="r"><code>con &lt;- readOrCreate(DataPath(&#39;ASD/con.rds&#39;), function() {
  mat &lt;- DataPath(&quot;AZ/cell_counts.csv&quot;) %&gt;% data.table::fread(sep=&quot;,&quot;) %&gt;%
    {set_rownames(mltools::sparsify(.[, 2:ncol(.)]), .$V1)} %&gt;%
    .[rowSums(. &gt; 0) &gt;= 10,]

  cell.metadata &lt;- DataPath(&quot;AZ/cell_metadata.csv&quot;) %&gt;% read_delim(delim=&#39;\t&#39;) %&gt;%
    rename(cell=sampleID, sample=patient) %&gt;%
    select(cell, batch, sample, sex, cellType, subclustID) %&gt;%
    filter(!grepl(&quot;un&quot;, sample), !(cellType %in% c(&#39;doublet&#39;, &#39;unID&#39;)))

  sample.metadata &lt;- group_by(cell.metadata, batch, sample, sex) %&gt;%
    summarise(n=n()) %&gt;% select(-n) %&gt;% lapply(setNames, .$sample)

  cell.metadata %&lt;&gt;% lapply(setNames, .$cell)

  mat.per.samp &lt;- splitMatrixByFactor(mat, cell.metadata$sample)

  dub.info &lt;- estimateDoubletInfo(mat.per.samp)
  p2s &lt;- plapply(mat.per.samp, createPagoda, min.transcripts.per.cell=500, dub.scores=dub.info$scores,
                 dub.threshold=0.3, mc.preschedule=TRUE, n.cores=N_CORES, progress=FALSE)

  createConos(p2s, sample.meta=sample.metadata, cell.meta=cell.metadata, n.cores=N_CORES)
}, force=FORCE) %&gt;% Conos$new()</code></pre>
<pre><code>.</code></pre>
<pre class="r"><code>con$plotGraph(groups=con$misc$cell_metadata$cellType, size=0.2, alpha=0.2)</code></pre>
<p><img src="figure/preprocess.Rmd/unnamed-chunk-2-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-2-1">
Past versions of unnamed-chunk-2-1.png
</button>
</p>
<div id="fig-unnamed-chunk-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
3085bf5
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>rm(con); gc();</code></pre>
<pre><code>           used  (Mb) gc trigger   (Mb)  max used   (Mb)
Ncells  4464430 238.5    8419650  449.7   8419650  449.7
Vcells 16905049 129.0  274219120 2092.2 535584051 4086.2</code></pre>
</div>
<div id="autism" class="section level2">
<h2>Autism</h2>
<p>Cells are already filtered by mitochondrial fraction of 0.05 and UMI threshold ~500, no additional filtration is needed.</p>
<pre class="r"><code>con &lt;- readOrCreate(DataPath(&#39;ASD/con.rds&#39;), function() {
  mat &lt;- Seurat::Read10X(DataPath(&quot;ASD&quot;)) %&gt;% .[rowSums(. &gt; 0) &gt;= 10,]
  meta &lt;- read_delim(DataPath(&quot;ASD/meta.txt&quot;), delim=&#39;\t&#39;) %&gt;%
    rename(cellType=cluster, PMI=`post-mortem interval (hours)`) %&gt;%
    mutate(cellType=gsub(&quot;-I(I)?&quot;, &quot;&quot;, cellType))

  sample.metadata &lt;- meta %&gt;%
    group_by(sample, individual, region, age, sex, diagnosis, Capbatch, Seqbatch) %&gt;%
    summarise(PMI=median(PMI)) %&gt;% lapply(setNames, .$sample)

  sample.metadata$region_hr &lt;- sample.metadata$sample %&gt;% strsplit(&#39;_&#39;) %&gt;% sapply(`[[`, 2)

  cell.metadata &lt;- meta %&gt;% lapply(setNames, .$cell)

  mat.per.samp &lt;- splitMatrixByFactor(mat, cell.metadata$sample)
  mat.per.cap &lt;- splitMatrixByFactor(mat, cell.metadata$Capbatch)

  dub.info &lt;- estimateDoubletInfo(mat.per.cap)
  p2s &lt;- plapply(mat.per.samp, createPagoda, dub.scores=dub.info$scores, dub.threshold=0.17,
                 mc.preschedule=TRUE, n.cores=N_CORES, progress=FALSE)
  createConos(p2s, sample.meta=sample.metadata, cell.meta=cell.metadata, n.cores=N_CORES)
}, force=FORCE) %&gt;% Conos$new()</code></pre>
<pre><code>.</code></pre>
<pre class="r"><code>con$plotGraph(groups=con$misc$cell_metadata$cellType, size=0.1, alpha=0.1)</code></pre>
<p><img src="figure/preprocess.Rmd/unnamed-chunk-4-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
3085bf5
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>rm(con); gc();</code></pre>
<pre><code>           used  (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells  5089547 271.9   15090983   806.0   32790919  1751.3
Vcells 93770050 715.5 3360870642 25641.5 3138941721 23948.3</code></pre>
</div>
<div id="epilepsy" class="section level2">
<h2>Epilepsy</h2>
<pre class="r"><code>con &lt;- readOrCreate(DataPath(&#39;EP/con.rds&#39;), function() {
  con.pap &lt;- read_rds(DataPath(&quot;EP/con_filt_samples.rds&quot;)) %&gt;% conos::Conos$new()
  cell.metadata &lt;- DataPath(&quot;EP/annotation.csv&quot;) %&gt;% read_csv() %&gt;%
    rename(cellType=l4) %&gt;% lapply(setNames, .$cell)
  sample.metadata &lt;- DataPath(&quot;EP/sample_info.csv&quot;) %&gt;% read_csv() %&gt;% lapply(setNames, .$Alias)

  con.pap$samples %&gt;% lapply(pagoda2::Pagoda2$new) %&gt;%
    createConos(sample.meta=sample.metadata, cell.meta=cell.metadata, k=40, n.cores=N_CORES)
}, force=FORCE) %&gt;% Conos$new()</code></pre>
<pre><code>.</code></pre>
<pre class="r"><code>con$plotGraph(groups=con$misc$cell_metadata$cellType, size=0.1, alpha=0.1, font.size=c(2, 3))</code></pre>
<p><img src="figure/preprocess.Rmd/unnamed-chunk-6-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-6-1">
Past versions of unnamed-chunk-6-1.png
</button>
</p>
<div id="fig-unnamed-chunk-6-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
3085bf5
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>rm(con); gc();</code></pre>
<pre><code>           used  (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells  5127848 273.9   15194756   811.5   36745925  1962.5
Vcells 88913614 678.4 2150957212 16410.6 3138941721 23948.3</code></pre>
</div>
<div id="ms" class="section level2">
<h2>MS</h2>
<p>The threshold on transcripts here is set only because there was 1 almost empty cell reported, all other cells already had enough transcripts.</p>
<pre class="r"><code>con &lt;- readOrCreate(DataPath(&#39;MS/con.rds&#39;), function() {
  mat &lt;- Seurat::Read10X(DataPath(&quot;MS&quot;)) %&gt;% .[rowSums(. &gt; 0) &gt;= 10,]
  meta &lt;- read_delim(DataPath(&quot;MS/meta.txt&quot;), delim=&#39;\t&#39;) %&gt;%
    mutate(cell_type=gsub(&quot;-(Cntl|MS(-1|-2)?)&quot;, &quot;&quot;, x=cell_type)) %&gt;%
    rename(cellType=cell_type)

  sample.metadata &lt;- meta[,5:14] %&gt;% split(.$sample) %&gt;% lapply(`[`, 1,) %&gt;%
    do.call(rbind, .) %&gt;% lapply(setNames, .$sample)
  cell.metadata &lt;- meta %&gt;% lapply(setNames, .$cell)

  mat.per.samp &lt;- splitMatrixByFactor(mat, cell.metadata$sample)
  mat.per.cap &lt;- splitMatrixByFactor(mat, cell.metadata$Capbatch)
  dub.info &lt;- estimateDoubletInfo(mat.per.cap)
  p2s &lt;- plapply(mat.per.samp, createPagoda, min.transcripts.per.cell=800, dub.scores=dub.info$scores,
                 dub.threshold=0.2, mc.preschedule=TRUE, n.cores=N_CORES, progress=FALSE)

  createConos(p2s, sample.meta=sample.metadata, cell.meta=cell.metadata, n.cores=N_CORES)
}, force=FORCE) %&gt;% Conos$new()</code></pre>
<pre><code>.</code></pre>
<pre class="r"><code>con$plotGraph(groups=con$misc$cell_metadata$cellType, size=0.1, alpha=0.1)</code></pre>
<p><img src="figure/preprocess.Rmd/unnamed-chunk-8-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
3085bf5
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>rm(con); gc();</code></pre>
<pre><code>           used  (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells  5078468 271.3   15194756   811.5   36745925  1962.5
Vcells 47645373 363.6 1720765770 13128.5 3138941721 23948.3</code></pre>
</div>
<div id="pf" class="section level2">
<h2>PF</h2>
<p>The paper already performed the filtration</p>
<pre class="r"><code>con &lt;- readOrCreate(DataPath(&#39;PF/con.rds&#39;), function() {
  cell.metadata &lt;- read_csv(DataPath(&quot;PF/cell_metadata.csv&quot;)) %&gt;% rename(cell=X1) %&gt;%
    filter(Diagnosis %in% c(&quot;Control&quot;, &quot;IPF&quot;)) %&gt;%
    rename(sample=Sample_Name, cellType=celltype)

  sample.metadata &lt;- cell.metadata %&gt;%
    group_by(sample, Sample_Source, Diagnosis, Status, orig.ident) %&gt;%
    summarise(n=n()) %&gt;% select(-n) %&gt;%
    lapply(setNames, .$sample)

  mat &lt;- DataPath(&quot;PF&quot;) %&gt;% Seurat::Read10X(gene.column=1) %&gt;%
    .[,cell.metadata$cell] %&gt;% .[rowSums(. &gt; 0) &gt;= 10,]

  cell.metadata %&lt;&gt;% lapply(setNames, .$cell)
  mat.per.samp &lt;- splitMatrixByFactor(mat, cell.metadata$sample)
  p2s &lt;- plapply(mat.per.samp, createPagoda, mc.preschedule=TRUE, n.cores=N_CORES, progress=FALSE)
  createConos(p2s, sample.meta=sample.metadata, cell.meta=cell.metadata, n.cores=N_CORES)
}, force=FORCE) %&gt;% Conos$new()</code></pre>
<pre><code>Warning: Missing column names filled in: &#39;X1&#39; [1]</code></pre>
<pre><code>
-- Column specification --------------------------------------------------------
cols(
  X1 = col_character(),
  orig.ident = col_character(),
  nCount_RNA = col_double(),
  nFeature_RNA = col_double(),
  Diagnosis = col_character(),
  Sample_Name = col_character(),
  Sample_Source = col_character(),
  Status = col_character(),
  percent.mt = col_double(),
  nCount_SCT = col_double(),
  nFeature_SCT = col_double(),
  seurat_clusters = col_double(),
  population = col_character(),
  celltype = col_character()
)</code></pre>
<pre><code>`summarise()` has grouped output by &#39;sample&#39;, &#39;Sample_Source&#39;, &#39;Diagnosis&#39;, &#39;Status&#39;. You can override using the `.groups` argument.</code></pre>
<pre><code>found 0 out of 231 cached PCA space pairs ... </code></pre>
<pre><code>running 231 additional PCA space pairs </code></pre>
<pre><code> done</code></pre>
<pre><code>inter-sample links using mNN </code></pre>
<pre><code> done</code></pre>
<pre><code>local pairs </code></pre>
<pre><code> done</code></pre>
<pre><code>building graph .</code></pre>
<pre><code>.</code></pre>
<pre><code>done</code></pre>
<pre><code>Warning in embedKnnGraph(commute.times, n.neighbors = n.neighbors, names = adj.info$names, : Maximal number of estimated neighbors is 27. Consider increasing min.visited.verts, min.prob or min.prob.lower.</code></pre>
<pre class="r"><code>con$plotGraph(groups=con$misc$cell_metadata$cellType, size=0.1, alpha=0.1)</code></pre>
<p><img src="figure/preprocess.Rmd/unnamed-chunk-10-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
3085bf5
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>rm(con); gc();</code></pre>
<pre><code>           used  (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells  5126196 273.8   15194756   811.5   36745925  1962.5
Vcells 52427576 400.0 1903262240 14520.8 3138941721 23948.3</code></pre>
</div>
<div id="scc" class="section level2">
<h2>SCC</h2>
<pre class="r"><code>con &lt;- readOrCreate(DataPath(&#39;SCC/con.rds&#39;), function() {
  mat &lt;- data.table::fread(DataPath(&quot;SCC/counts.txt&quot;), sep=&quot;\t&quot;) %&gt;%
    {set_rownames(mltools::sparsify(.[3:nrow(.), 2:ncol(.)]), .$V1[3:nrow(.)])}

  cell.metadata &lt;- read_delim(DataPath(&#39;SCC/cell_metadata.txt&#39;), delim=&#39;\t&#39;) %&gt;%
    filter(!(level3_celltype %in% c(&#39;Multiplet&#39;, &#39;Keratinocyte&#39;))) %&gt;%
    rename(cell=nCount_RNA, cellType=level3_celltype)
  cell.metadata$sample &lt;- cell.metadata$cell %&gt;% strsplit(&quot;_&quot;) %&gt;%
    sapply(function(x) paste(x[1:2], collapse=&#39;_&#39;))
  cell.metadata$cellType %&lt;&gt;% gsub(&quot;(Normal|Tumor)_&quot;, &quot;&quot;, .)
  cell.metadata %&lt;&gt;% lapply(setNames, .$cell)

  mat &lt;- mat[rowSums(mat &gt; 0) &gt;= 10, cell.metadata$cell]
  mat.per.samp &lt;- splitMatrixByFactor(mat, cell.metadata$sample)

  p2s &lt;- plapply(mat.per.samp, createPagoda, min.transcripts.per.cell=800,
                mc.preschedule=TRUE, n.cores=N_CORES, progress=FALSE)

  createConos(p2s, sample.meta=NULL, cell.meta=cell.metadata, n.cores=N_CORES)
}, force=FORCE) %&gt;% Conos$new()</code></pre>
<pre><code>Warning in data.table::fread(DataPath(&quot;SCC/counts.txt&quot;), sep = &quot;\t&quot;): Detected
48164 column names but the data has 48165 columns (i.e. invalid file). Added 1
extra default column name for the first column which is guessed to be row names
or an index. Use setnames() afterwards if this guess is not correct, or fix the
file write command that created the file to create a valid file.</code></pre>
<pre><code>
-- Column specification --------------------------------------------------------
cols(
  nCount_RNA = col_character(),
  nFeature_RNA = col_double(),
  patient = col_double(),
  tum.norm = col_character(),
  level1_celltype = col_character(),
  level2_celltype = col_character(),
  level3_celltype = col_character()
)</code></pre>
<pre><code>Warning: 48164 parsing failures.
row col  expected    actual                                                                                file
  1  -- 7 columns 8 columns &#39;/d0-mendel/home/viktor_petukhov/projects/cacoaAnalysis/data/SCC/cell_metadata.txt&#39;
  2  -- 7 columns 8 columns &#39;/d0-mendel/home/viktor_petukhov/projects/cacoaAnalysis/data/SCC/cell_metadata.txt&#39;
  3  -- 7 columns 8 columns &#39;/d0-mendel/home/viktor_petukhov/projects/cacoaAnalysis/data/SCC/cell_metadata.txt&#39;
  4  -- 7 columns 8 columns &#39;/d0-mendel/home/viktor_petukhov/projects/cacoaAnalysis/data/SCC/cell_metadata.txt&#39;
  5  -- 7 columns 8 columns &#39;/d0-mendel/home/viktor_petukhov/projects/cacoaAnalysis/data/SCC/cell_metadata.txt&#39;
... ... ......... ......... ...................................................................................
See problems(...) for more details.</code></pre>
<pre><code>found 0 out of 190 cached PCA space pairs ... </code></pre>
<pre><code>running 190 additional PCA space pairs </code></pre>
<pre><code> done</code></pre>
<pre><code>inter-sample links using mNN </code></pre>
<pre><code> done</code></pre>
<pre><code>local pairs </code></pre>
<pre><code> done</code></pre>
<pre><code>building graph .</code></pre>
<pre><code>.</code></pre>
<pre><code>done</code></pre>
<pre><code>Warning in embedKnnGraph(commute.times, n.neighbors = n.neighbors, names = adj.info$names, : Maximal number of estimated neighbors is 39. Consider increasing min.visited.verts, min.prob or min.prob.lower.</code></pre>
<pre class="r"><code>con$plotGraph(groups=con$misc$cell_metadata$cellType, size=0.1, alpha=0.1)</code></pre>
<p><img src="figure/preprocess.Rmd/unnamed-chunk-12-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-1">
Past versions of unnamed-chunk-12-1.png
</button>
</p>
<div id="fig-unnamed-chunk-12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
3085bf5
</td>
<td>
viktor_petukhov
</td>
<td>
2021-11-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>rm(con); gc();</code></pre>
<pre><code>            used  (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells   5182595 276.8   15194756   811.5   36745925  1962.5
Vcells 122581181 935.3 2381939596 18172.8 5813473782 44353.3</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.6 LTS

Matrix products: default
BLAS:   /usr/local/R/R-4.0.3/lib/R/lib/libRblas.so
LAPACK: /usr/local/R/R-4.0.3/lib/R/lib/libRlapack.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] cacoaAnalysis_0.1.0 reticulate_1.22     dataorganizer_0.1.0
 [4] conos_1.4.4         pagoda2_1.0.7       igraph_1.2.6       
 [7] Matrix_1.2-18       sccore_1.0.0        magrittr_2.0.1     
[10] forcats_0.5.1       stringr_1.4.0       dplyr_1.0.7        
[13] purrr_0.3.4         readr_1.4.0         tidyr_1.1.4        
[16] tibble_3.1.5        ggplot2_3.3.5       tidyverse_1.3.0    
[19] workflowr_1.6.2    

loaded via a namespace (and not attached):
  [1] utf8_1.2.2           R.utils_2.10.1       tidyselect_1.1.1    
  [4] htmlwidgets_1.5.3    grid_4.0.3           Rtsne_0.15          
  [7] devtools_2.3.2       munsell_0.5.0        ica_1.0-2           
 [10] codetools_0.2-16     miniUI_0.1.1.1       future_1.22.1       
 [13] withr_2.4.2          colorspace_2.0-2     highr_0.9           
 [16] knitr_1.36           rstudioapi_0.13      Seurat_4.0.0        
 [19] stats4_4.0.3         ROCR_1.0-11          tensor_1.5          
 [22] listenv_0.8.0        labeling_0.4.2       git2r_0.27.1        
 [25] urltools_1.7.3       polyclip_1.10-0      farver_2.1.0        
 [28] rprojroot_2.0.2      parallelly_1.28.1    Matrix.utils_0.9.8  
 [31] vctrs_0.3.8          generics_0.1.0       xfun_0.26           
 [34] R6_2.5.1             doParallel_1.0.16    ggbeeswarm_0.6.0    
 [37] clue_0.3-59          spatstat.utils_2.0-0 cachem_1.0.6        
 [40] assertthat_0.2.1     promises_1.1.1       scales_1.1.1        
 [43] beeswarm_0.4.0       gtable_0.3.0         globals_0.14.0      
 [46] goftest_1.2-2        processx_3.4.5       drat_0.1.8          
 [49] rlang_0.4.11         GlobalOptions_0.1.2  splines_4.0.3       
 [52] lazyeval_0.2.2       broom_0.7.9          brew_1.0-6          
 [55] abind_1.4-5          yaml_2.2.1           reshape2_1.4.4      
 [58] modelr_0.1.8         backports_1.2.1      httpuv_1.5.4        
 [61] tools_4.0.3          usethis_1.6.3        ellipsis_0.3.2      
 [64] jquerylib_0.1.4      RColorBrewer_1.1-2   BiocGenerics_0.36.1 
 [67] ggridges_0.5.3       sessioninfo_1.1.1    Rcpp_1.0.7          
 [70] plyr_1.8.6           ps_1.4.0             prettyunits_1.1.1   
 [73] deldir_0.2-10        rpart_4.1-15         dendsort_0.3.3      
 [76] pbapply_1.4-3        GetoptLong_1.0.5     cowplot_1.1.1       
 [79] zoo_1.8-8            S4Vectors_0.28.1     SeuratObject_4.0.0  
 [82] grr_0.9.5            haven_2.4.1          ggrepel_0.9.1       
 [85] cluster_2.1.0        fs_1.5.0             here_1.0.1          
 [88] scattermore_0.7      data.table_1.14.2    RSpectra_0.16-0     
 [91] circlize_0.4.13      lmtest_0.9-38        triebeard_0.3.0     
 [94] reprex_0.3.0         RANN_2.6.1           whisker_0.4         
 [97] fitdistrplus_1.1-3   matrixStats_0.61.0   pkgload_1.2.1       
[100] patchwork_1.1.1      xtable_1.8-4         mime_0.12           
[103] hms_1.1.1            evaluate_0.14        RMTstat_0.3         
[106] N2R_0.1.1            readxl_1.3.1         IRanges_2.24.1      
[109] gridExtra_2.3        shape_1.4.6          testthat_3.0.0      
[112] compiler_4.0.3       KernSmooth_2.23-17   crayon_1.4.1        
[115] R.oo_1.24.0          htmltools_0.5.2      mgcv_1.8-33         
[118] later_1.1.0.1        lubridate_1.7.9.2    DBI_1.1.1           
[121] dbplyr_2.0.0         ComplexHeatmap_2.9.4 MASS_7.3-53         
[124] rappdirs_0.3.3       cli_3.0.1            R.methodsS3_1.8.1   
[127] parallel_4.0.3       pkgconfig_2.0.3      plotly_4.9.3        
[130] xml2_1.3.2           foreach_1.5.1        vipor_0.4.5         
[133] leidenAlg_0.1.0      rvest_0.3.6          callr_3.5.1         
[136] digest_0.6.28        sctransform_0.3.2    RcppAnnoy_0.0.18    
[139] spatstat.data_2.0-0  leiden_0.3.7         rmarkdown_2.11      
[142] cellranger_1.1.0     Rook_1.1-1           uwot_0.1.10         
[145] shiny_1.5.0          rjson_0.2.20         lifecycle_1.0.1     
[148] nlme_3.1-149         mltools_0.3.5        jsonlite_1.7.2      
[151] viridisLite_0.4.0    desc_1.3.0           fansi_0.5.0         
[154] pillar_1.6.3         lattice_0.20-41      ggrastr_0.2.1       
[157] fastmap_1.1.0        httr_1.4.2           pkgbuild_1.1.0      
[160] survival_3.2-7       glue_1.4.2           remotes_2.2.0       
[163] spatstat_1.64-1      png_0.1-7            iterators_1.0.13    
[166] stringi_1.7.5        memoise_2.0.0        irlba_2.3.3         
[169] future.apply_1.8.1  </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
